### PR TITLE
build: Auto-gen tsec tsconfig files

### PR DESCRIPTION
**Pull request by uraj**
Previously, tsec-related tsconfig.json files are manually managed.
Switch to automatically generated ones.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [X] Build related changes
- [X] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by uraj** [comment]
@josephperrott Here's the patch the removes all the manually maintained tsconfig-tsec.json files and generate them within bazel rules instead.

**Comment by uraj** [comment]
@devversion this should also make it easier to make `tsec_test` work on Windows. Could you take a look?

**Review by josephperrott — COMMENTED** [review]
LGTM

We should wait for @devversion to confirm about how this helps progress towards usage on Windows and his review of the change as well before merging

File: tools/tsec.bzl [hunk 1]
```diff
@@ -1,107 +1,141 @@
 """Bazel rules and macros for running tsec over a ng_module or ts_library."""
 
+load("@bazel_skylib//lib:new_sets.bzl", "sets")
 load("@npm//@bazel/typescript/internal:ts_config.bzl", "TsConfigInfo")
 load("@build_bazel_rules_nodejs//:providers.bzl", "DeclarationInfo")
 load("@npm//tsec:index.bzl", _tsec_test = "tsec_test")
 
-TsecTsconfigInfo = provider(fields = ["src", "exemption", "deps"])
+TsLibInfo = provider(fields = ["srcs", "deps", "module_name", "paths"])
```
Comment (devversion) [inline tools/tsec.bzl:8]: Can we name this more specific to tsec. It's obviously clear that this is not related to TypeScript's `tslib` package, but it shouldn't hurt naming this `TsecTargetInfo` or something similar.

File: tools/tsec.bzl [hunk 2]
```diff
@@ -1,107 +1,141 @@
...
+
+    if base_tsconfig:
+        tsconfig["extends"] = base_url + "/" + base_tsconfig.path
+
+    compiler_options = {"noEmit": True}
+    compiler_options["baseUrl"] = base_url
+
+    tslib_info = target[TsLibInfo]
+    paths = dict(**tslib_info.paths)
+    paths.pop(tslib_info.module_name)
```
Comment (devversion) [inline tools/tsec.bzl:49]: Nit: add a comment that we omit the path mapping for the target that is actually built. Although in practice, keeping this should not cause any issues, right?

File: tools/tsec.bzl [hunk 3]
```diff
@@ -1,107 +1,141 @@
...
+    deps = []
+
+    base_tsconfig_src = None
+    base = ctx.attr.base
+
+    # Gather all base tsconfig files and the exemption list.
+    if base:
+        if TsConfigInfo not in base:
+            fail("`base` must be a ts_config target")
+        deps.append(depset(base[TsConfigInfo].deps))
```
Comment (devversion) [inline tools/tsec.bzl:91]: Nit: I think we don't necessarily need to put this into a depset if we can just concatenate it later with the `ctx.outputs.out` value.

**Review by devversion — APPROVED** [review]
LGTM 👍 

Yeah, the auto-generating will definitely help if we want to setup the various `rootDirs` to support execroot and bazel-bin file resolution on windows. It's generally much nicer that way IMO.

File: tools/tsec.bzl (hunk 3 above)
Comment (uraj) [reply tools/tsec.bzl:91]: Done.

File: tools/tsec.bzl (hunk 2 above)
Comment (uraj) [reply tools/tsec.bzl:49]: Yes, it doesn't hurt to keep it.

File: tools/tsec.bzl (hunk 1 above)
Comment (uraj) [reply tools/tsec.bzl:8]: Done.

**Review by devversion — APPROVED** [review]

**Comment by jessicajaniuk** [comment]
Looks like there's conflicts with `12.2.x`. I'm updating the target to minor to get this landed in master. Can you create a patch PR targeting `12.2.x`?

**Comment by uraj** [comment]
@jessicajaniuk do you still need a PR targeting `12.2.x`?

**Comment by devversion** [comment]
@uraj Should be fine without an extra PR for patch (12.2.x). The main PR did not land in the patch branch either

**Pull request closed without merging**